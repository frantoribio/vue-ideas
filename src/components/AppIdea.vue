<template>
  <article class="rounded-lg bg-gray-300 mb-4 p-3 sm:(flex items-center) ">
    <!-- Info -->
    <section class="text-center sm:(flex-1 text-left) ">
      <h2 class="text-xl sm:(leading-6 text-2xl) ">
        {{ idea.name }}
      </h2>
      <small>{{ idea.userName }}
        <i>, {{ timeAgo(idea.createdAt) }}</i>
      </small>
    </section>
    <!-- Votos -->
    <section class="border-black border-t-2 mt-6 pt-3 sm:(pt-0 pl-3 border-t-0 border-l-2 mt-0 flex items-center) ">
      <h3
        id="votes"
        class="font-bold text-center text-3xl"
      >
        {{ idea.votes }}
      </h3>
      <nav
        v-if="user && !userVoted"
        class="flex justify-center sm:(block pl-2)"
      >
        <img
          id="voteup"
          class="cursor-pointer w-10"
          src="../assets/images/arrow.svg"
          alt="Vote up"
          @click="voteIdea(true)"
        >
        <img
          id="votedown"
          class="cursor-pointer transform w-10 -rotate-180"
          src="../assets/images/arrow.svg"
          alt="Vote down"
          @click="voteIdea(false)"
        >
      </nav>
    </section>
  </article>
</template>

<script>
import { computed } from 'vue'
import dayjs from 'dayjs'
import 'dayjs/locale/es'
import relativeTime from 'dayjs/plugin/relativeTime'

dayjs.locale('es')
dayjs.extend(relativeTime)

export default {
  name: 'AppIdea',
  props: {
    idea: {
      type: Object,
      required: true
    },
    user: {
      type: [Object, null]
    }
  },
  emits: ['vote-idea'],

  setup (props, { emit }) {
    const voteIdea = (type) => emit('vote-idea', { id: props.idea.id, type })
    const userVoted = computed(() => {
      if (props.user.votes) {
        return props.user.votes.find(item => item === props.idea.id)
      } else {
        return false
      }
    })

    const timeAgo = (timestamp) => {
      if (timestamp) {
        const date = new Date(timestamp)
        return dayjs().from(dayjs(date), true)
        // return dayjs(date).format('DD/MM/YYYY HH:mm:ss')
      }
      return ''
    }

    return { voteIdea, userVoted, timeAgo }
  }
}
</script>

<style lang="scss" scoped>

</style>
