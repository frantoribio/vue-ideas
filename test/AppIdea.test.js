import { mount } from '@vue/test-utils'
import { expect, test } from 'vitest'
import AppIdea from '../src/components/AppIdea.vue'

test('mount component', async () => {
  const idea = {
    name: 'Idea name 1',
    userName: 'Username 1',
    votes: 99
  }

  expect(AppIdea).toBeTruthy()
  const wrapper = mount(AppIdea, {
    props: {
      idea: idea
    }
  })
  expect(wrapper.text()).toContain('Idea name 1')
  expect(wrapper.text()).toContain('Username 1')
  expect(wrapper.html()).toMatchSnapshot()
  const votes = wrapper.find('#votes').text().trim()
  expect(votes).toBe('99')
  const up = wrapper.find('#voteup')
  await up.trigger('click')
  expect(wrapper.emitted('vote-idea')).toBeTruthy()
  // const event = wrapper.emitted('vote-idea')[0][0]
  // expect(event.type).toBe(true)
  expect(wrapper.emitted('vote-idea').length).toBe(1)
  const down = wrapper.find('#votedown')
  await down.trigger('click')
  expect(wrapper.emitted('vote-idea').length).toBe(2)
})
