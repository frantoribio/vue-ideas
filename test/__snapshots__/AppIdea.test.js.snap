// Vitest Snapshot v1

exports[`mount component 1`] = `
"<article class=\\"rounded-lg bg-gray-300 mb-4 p-3 sm:flex sm:items-center\\">
  <!-- Info -->
  <section class=\\"text-center sm:flex-1 sm:text-left\\">
    <h2 class=\\"text-xl sm:leading-6 sm:text-2xl\\">Idea name 1</h2><small>by: Username 1</small>
  </section><!-- Votos -->
  <section class=\\"border-black border-t-2 mt-6 pt-3 sm:pt-0 sm:pl-3 sm:border-t-0 sm:border-l-2 sm:mt-0 sm:flex sm:items-center\\">
    <h3 id=\\"votes\\" class=\\"font-bold text-center text-3xl\\">99</h3>
    <nav class=\\"flex justify-center sm:block sm:pl-2\\"><img id=\\"voteup\\" class=\\"cursor-pointer w-10\\" src=\\"/src/assets/images/arrow.svg\\" alt=\\"Vote up\\" /><img id=\\"votedown\\" class=\\"cursor-pointer transform w-10 -rotate-180\\" src=\\"/src/assets/images/arrow.svg\\" alt=\\"Vote down\\" /></nav>
  </section>
</article>"
`;
